### ISSUE TITLE
Could ruff also check files according to file's shebang?

### ISSUE THREAD

**Issue by praiskup** [astral-sh/ruff#13122]
```
$ ruff  check .
warning: No Python files found under the given path(s)
All checks passed!
$ head -1 vcs-diff-lint
#! /usr/bin/python3
$ ruff check vcs-diff-lint
All checks passed!
```

Even though the file is not `*.py`, it still is a Python file and it is worth checking.

**Comment by dhruvmanila** [comment]
Related https://github.com/astral-sh/ruff/issues/2192

**tl;dr** It's recommended to use the `include` / `extend-include` config options to explicitly list down the files that you want Ruff to check for. Adding support for it _might_ come with a performance cost although not sure.

**Comment by dhruvmanila** [comment]
Does using [`include`](https://docs.astral.sh/ruff/settings/#include) / [`extend-include`](https://docs.astral.sh/ruff/settings/#extend-include) solve your use case? If so, we can close this issue.

**Comment by praiskup** [comment]
The differential analysis tool we develop uses Ruff as one of the backends, and while we appreciate Ruff's overall performance - analyzing the shebang is worth it in our case.

The goal is to make future project enablements (= "start using vcs-diff-lint + ruff" in more projects) as trivial as possible.  Explicitly asking for `include/extend-include` per/project raises the initial enablement barrier, and incomplete analysis is also not nice...

That said, having an `analyze-shebangs = true` opt-in would be completely fine for our use case; we can afford to spend some time on this.

**Comment by praiskup** [comment]
https://github.com/fedora-copr/vcs-diff-lint/pull/25 seems to help a bit here;  vcs-diff-lint knows the set of **changed** Python, including those that do not end with `*.py`.  So we can indeed use `extend-include` to at least cover those changed files, but we are still not checking unchanged non-`*.py` files.

**Comment by MichaReiser** [comment]
Detecting all Python files based on shebang has the downside that Ruff needs to read at least the first 80 or so bytes to determine whether the file should be included in the analysis. This is significantly more costly than just looking at a file extension, so I would not want this behavior to be the default. I'm open to adding such a feature if we can come up with an interface that doesn't penalty the default use case where this isn't needed.

**Comment by praiskup** [comment]
> Detecting all Python files based on shebang has the downside that Ruff needs to read at least the first 80 or so

Could we, e.g., only include files with no "suffixes"?  That would be the most common case, I assume.

Otherwise, opt-in is completely fine!

**Comment by pmp-p** [comment]
Also some html files - though valid python - have a non standard shebang (like it used to be on MS-DOS) that Ruff cannot yet support/handle ( unlike black -x / python -x )

eg source : https://github.com/pygame-web/showroom/blob/main/pygame-scripts/org.pygame.touchpong.html

for run with python-wasm : https://pygame-web.github.io/showroom/pygame-scripts/org.pygame.touchpong.html

**Comment by praiskup** [comment]
> Also some html files - though valid python - have a non standard shebang

Indeed, this should be only an additional/optional analysis, and I don't think we need to have 100% success rate (nb. the basic `.py` suffix rule isn't 100% correct either).

**Comment by daethnir** [comment]
Adding another +1 to supporting shebang lines.

Our executable scripts do not end in `.py` and have `chmod +`. In most cases they are `#!/usr/bin/env python3`.

However in some cases we have wrappers ala `#!/usr/bin/env -S uvwrapper ...` that automatically run the tool through `uv` - so supporting user-specifiable shebang lines would be even better.

**Comment by MichaReiser** [comment]
@daethnir I don't know how many executable scripts you have but if there are only a few, does the solution outlined by @dhruvmanila might work for you in the meantime:

> tl;dr It's recommended to use the include / extend-include config options to explicitly list down the files that you want Ruff to check for. Adding support for it might come with a performance cost although not sure.

**Comment by MichaReiser** [comment]
Doing the shebang check only for files without extension might not be that expensive, as they shouldn't be very common.

**Comment by daethnir** [comment]
> [@daethnir](https://github.com/daethnir) I don't know how many executable scripts you have

It's not the quantity - it's assuring we don't miss any. If we need to manually add them no include lists then it's guaranteed we'll miss at some point.

Right now for those places we're using pre-commit we've added a second hashish hook w/ code like this:


```bash
...
for file in "$@"; do
    # Skip any files with an extension (any dot of any kind - could have false negatives)
    if echo "$file" | grep -q '\.[^/]' ; then
        continue
    fi

    # Process any files that are uv / python
    if head -1 "$file" | grep -q -E '^#!.*\benv\b.*uv|^#!.*python' ; then
        ruff check --fix "$file" || success=false
        ruff format "$file" || success=false
    fi
done
```

This "works" because in pre-commit it will force feed us a list of files to check (the `"$@"`) but would be much nicer if we could not need to add hackish hooks and supporting scripts to each repo - this method has a few drawbacks

* can't just run `ruff` with whatever args you want
* need to maintain hooks in two places now and keep in sync

> tl;dr It's recommended to use the include / extend-include config options to explicitly list down the files that you want Ruff to check for. Adding support for it might come with a performance cost although not sure.

If ruff only checked files that

* are executable (stat)
* have python / uv in the first line after a shebang

I bet it's more performant than one would expect. Folks don't chmod non-scripts for funzies.

**Comment by daethnir** [comment]
> Doing the shebang check only for files without extension might not be that expensive, as they shouldn't be very common.

No extension (already known by ruff's check), and executable (rare), and read ~40 characters. Probably not a big lift.
